openedx/openedx-platform
# One student module history per request

NOTE: THIS SOLUTION IS DEPENDENT ON #31189

## Description

When a student submits a problem answer, the state is stored in a StudentModule record containing answer, score, correctness, etc. The record, though, is updated in multiple steps within the single request (first the grade is updated, then the state is updated separately). Each partial save would trigger a separate StudentModuleHistory record to be stored resulting in duplicate and inaccurate historical records.

This solution uses the RequestCache to track within a request thread which StudentModules are updated and a single corresponding StudentModuleHistory id. If multiple update actions occur within the request cycle, then modify the history record that was already generated to ensure that each submission only results in one StudentModuleHistory record.

## Supporting information

This issue and its solution were discussed [here](https://discuss.openedx.org/t/extra-history-record-stored-on-each-problem-submission/8081)

## Testing instructions

Submit an answer for any problem xblock, then confirm that the `courseware_studentmodulehistory` table only contains one new row corresponding to the latest submission.

## Deadline

None

## Other information

None
